### ISSUE TITLE
[MPMD-174] Using a permalink from Sonar as a ruleset does not work

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-pmd-plugin#523]
**[Cremers stijn](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cremersstijn)** opened **[MPMD-174](https://issues.apache.org/jira/browse/MPMD-174?redirect=false)** and commented

I am trying to use a permalink from sonar with the pmd configuration from sonar as a rulest in the maven-pmd-plugin.

This is my maven configuration:
\<plugin>
\<groupId>org.apache.maven.plugins\</groupId>
\<artifactId>maven-pmd-plugin\</artifactId>
\<version>3.0.1\</version>
\<configuration>
\<rulesets>
\<ruleset>http://my-tools.mycompany.com/sonar/profiles/export?format=pmd&amp;language=java&amp;name=MyProfile\</ruleset>
\</rulesets>
\</configuration>
\<executions>
\<execution>
\<id>pmd\</id>
\<goals>\<goal>check\</goal>\</goals>
\</execution>
\</executions>
\</plugin>

But i get the following error when i run "mvn clean install":
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.0.1:pmd (pmd) on project my-pmd:
An error has occurred in PMD Report report generation. Could not find resource 'rulesets/http://my/tools.mycompany.com/sonar/profiles.xml'. -> [Help 1]

I have tried to use the link from the sonar demo site:
\<plugin>
\<groupId>org.apache.maven.plugins\</groupId>
\<artifactId>maven-pmd-plugin\</artifactId>
\<version>3.0.1\</version>
\<configuration>
\<rulesets>
\<ruleset>http://nemo.sonarqube.org/profiles/export?format=pmd&amp;language=java&amp;name=Nemo\</ruleset>
\</rulesets>
\</configuration>
\<executions>
\<execution>
\<id>pmd\</id>
\<goals>\<goal>check\</goal>\</goals>
\</execution>
\</executions>
\</plugin>

But then i get the following error:
[INFO] --- maven-pmd-plugin:3.0.1:pmd (pmd) @ my-pmd ---
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] Failure executing PMD: Couldn't find the class White spaces are required between publicId and systemId.
java.lang.RuntimeException: Couldn't find the class White spaces are required between publicId and systemId.
at net.sourceforge.pmd.RuleSetFactory.classNotFoundProblem(RuleSetFactory.java:244)
at net.sourceforge.pmd.RuleSetFactory.parseRuleSetNode(RuleSetFactory.java:238)
at net.sourceforge.pmd.RuleSetFactory.createRuleSet(RuleSetFactory.java:161)
at net.sourceforge.pmd.RuleSetFactory.createRuleSets(RuleSetFactory.java:126)
at net.sourceforge.pmd.RuleSetFactory.createRuleSets(RuleSetFactory.java:111)
at net.sourceforge.pmd.processor.AbstractPMDProcessor.createRuleSets(AbstractPMDProcessor.java:56)
at net.sourceforge.pmd.processor.MonoThreadProcessor.processFiles(MonoThreadProcessor.java:41)
at net.sourceforge.pmd.PMD.processFiles(PMD.java:271)
at org.apache.maven.plugin.pmd.PmdReport.generateReport(PmdReport.java:296)
at org.apache.maven.plugin.pmd.PmdReport.execute(PmdReport.java:194)
at org.apache.maven.plugin.pmd.PmdReport.executeReport(PmdReport.java:168)
at org.apache.maven.reporting.AbstractMavenReport.generate(AbstractMavenReport.java:190)
at org.apache.maven.reporting.AbstractMavenReport.execute(AbstractMavenReport.java:99)
at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:106)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
at org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions(MojoExecutor.java:364)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:198)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:318)
at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:153)
at org.apache.maven.cli.MavenCli.execute(MavenCli.java:555)
at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:214)
at org.apache.maven.cli.MavenCli.main(MavenCli.java:158)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
at java.lang.reflect.Method.invoke(Method.java:597)
at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:414)
at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:357)

---

**Attachments:**
- [MPMD-174.patch](https://issues.apache.org/jira/secure/attachment/12717194/MPMD-174.patch) (_12.75 kB_)
- [MPMD-174.patch](https://issues.apache.org/jira/secure/attachment/12717251/MPMD-174.patch) (_1.27 kB_)
- [MPMD-174-2.patch](https://issues.apache.org/jira/secure/attachment/12717267/MPMD-174-2.patch) (_11.57 kB_)
- [MPMD-174-3.patch](https://issues.apache.org/jira/secure/attachment/12717309/MPMD-174-3.patch) (_11.73 kB_)
- [MPMD-174-4.patch](https://issues.apache.org/jira/secure/attachment/12717289/MPMD-174-4.patch) (_8.29 kB_)
- [Sonar's main page.jpg](https://issues.apache.org/jira/secure/attachment/12717219/Sonar%27s+main+page.jpg) (_47.42 kB_)
- [test.log](https://issues.apache.org/jira/secure/attachment/12717204/test.log) (_37.54 kB_)

**Issue Links:**
- [MPMD-198](https://issues.apache.org/jira/browse/MPMD-198) Upgrade to PMD 5.2.3 as soon as it is available
 (_**"depends upon"**_)

9 votes, 10 watchers

**Comment by jira-importer** [comment]
**[Marek Siller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chearius)** commented

I can confirm this issue as well. Everything is working fine as long as you use a local ruleset. As soon as you switch to ruleset from SonarQube, it doesn't work.

Essentially what happens is following: maven-pmd-plugin calls PMD -> it creates new instance of RuleSetReferenceId. In the constructor, several assumptions are made, so eventually it strips everything after the last "/" from the URL (in case of SonarQube http://sonar-instance/profiles). This URL gets passed back to maven-pmd-plugin for download. It downloads a HTML page and stores it in a temporary file. This file gets passed to PMD as a ruleset and you get an exception, because the downloaded HTML page is not a PMD ruleset.

The problem can be solved in three ways:
* in maven-pmd-plugin: see attached patch file. If a http or https URL is specified, it just downloads the ruleset and doesn't parse the URL using RuleSetReferenceId constructor.
* as a workaround: just add ".xml" to your profiles' names in SonarQube
* in PMD: fix the constructor of RuleSetReferenceId so that it can handle HTTP URLs

I currently use a patched version of maven-pmd-plugin as I don't want to rename my profiles in SonarQube.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I know SonarQube only by name, so please fill in affected version and component to triage this bug.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Here is a view of sonar's main page.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Here is the fix that  Marek Siller proposed plus a unit test that will actually test the failure.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Can we get this committed and a version released please?

Right now my PMD is broken.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks for the updated patch. I am a little bit confused by the statement

> The problem is actually in PMD's code (in RuleSetReferenceId class), where constructor builds rule set filename.

Is the problem located in the plugin or PMD itself?

I'll have a look at your patch this weekend.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

I don't know if it is a bug in the PMD's code or that's by design as I don't know anything about it. However, the root cause of this bug is the way that RuleSetReferenceId class processes the filename and when that happens to be a HTTP(S) URL it strips is it incorrectly. Please note that you could make similar patch in PMD: if the given ID is HTTP(S) just use the whole thing as ID and don't do any stripping.

Marek's explanation is dead on:

"Essentially what happens is following: maven-pmd-plugin calls PMD -> it creates new instance of RuleSetReferenceId. In the constructor, several assumptions are made, so eventually it strips everything after the last "/" from the URL (in case of SonarQube http://sonar-instance/profiles). This URL gets passed back to maven-pmd-plugin for download. It downloads a HTML page and stores it in a temporary file. This file gets passed to PMD as a ruleset and you get an exception, because the downloaded HTML page is not a PMD ruleset."

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

If this issue can be directly fixed in PMD, then that should be done there. Andreas Dangel has been very quick lately on fixing bugs. Please [file an issue](http://sourceforge.net/p/pmd/bugs/) with PMD and link to this one. PMD 5.2.2 is pending and chances are good that we get the fix real fast.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

As requested I created a defect in PMD:

https://sourceforge.net/p/pmd/bugs/1290/

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Hi Michael,

The fix has been merged into PMD's code. Unfortunately it will need to wait until 5.2.3 gets released. Can you update plugins pom to new PMD version when that gets released?

Regards,
Maciek

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Can you try PMD from master against this plugin from trunk whether everything works fine without further code modification?

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Hi Michael,

Sorry for the delay. So I verified that the master PMD branch 5.2.3-SNAPSHOT works with the trunk plugin. If I were you I would still merge the unit (integration) test you can catch any future regression.

Regards,
Maciek

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will do that as soon as the update of PMD has happened.

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

I just released PMD 5.2.3 which includes the bugfix for this issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This patch does not apply. See attached log file. [Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk), please change the patch.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Thanks Andreas :)

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Hi Michael,

Looks like that mock HTTP server causes all kind of issues with SLF4J version. This is partially because you are using ancient version of maven-plugin-testing-harness which depends on pre 1.6.x SLF4J. Also another issue was that Andreas changed my implementation little bit and it now uses HEAD request type, which mock server does not support.

For the latter reason I changed the mocking HTTP server framework to WireMock which does support HEAD requests. Please note that I still couldn't use the Rule mock as with such a old maven-plugin-testing-harness the `@Rule` annotation does not work.

I would create an issue about updating that plugin to something more recent. The current version is in 3.x range.

Anyways, here is a new patch. You can try with PMD 5.2.2 and you should see unit tests fail with same error as this issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Just a stupid question: Why not use Jetty for that? I think MPIR does that and we can steal from.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

WireMock is specifically made for testing so the integration is great and syntax is pretty fluent. You could probably rewrite tests, but you need to be able to serve different content based on the requesting URL so keep that in mind.

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

I've recreated the patch (see MPMD-174-4.patch) and changed the following:
* slf4j is excluded from wiremock rather than maven-plugin-testing-harness. It seems to be working.
* Patch file includes the new files only once (Maciek / Michael: are you using tortoise svn? I've seen this behavior earlier...)

For me all unit tests and integration tests are successful.

Btw. - Wiremock uses internally Jetty. It's just a layer on top to make it easier to configure jetty for unit testing.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [r1647545](http://svn.apache.org/r1647545). Guys, thank you for your help.

Andreas, I use all of them: `svn`, TortoiseSVN and Subclipse.

**Comment by jira-importer** [comment]
**[Maciek Siemczyk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msiemczyk)** commented

Thanks guys.

I didn't realize the new files were added twice. Yes, I am using TortoiseSVN to create patch,

**Comment by jira-importer** [comment]
**[Jochen Ehret](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jochenehret)** commented

Can you please tell us when the maven-pmd-plugin 3.4 will be released and available on Maven Central?

We've just migrated to SonarQube 4.5.1 and now lots of users get the "[WARNING] Failure executing PMD" problem mentioned above, so we really need to update.

Thanks and Best Regards,

Jochen.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Hi Jochen, my intention was to wait for the patch for MPMD-199 and then start a release. You are welcome to improve the patch in that ticket.
